### ISSUE TITLE
Id of WAL operation can not be duplicated

### ISSUE THREAD

**Issue by PhantomYdn** [orientechnologies/orientdb#9568]
## Expected behavior  
After successful shutdown, startup should be also successful. Actually it should be successful in any case.

## Actual behavior  
Startup is started to fail with the following exception:

```
2021-03-20 20:11:18:631 INFO  Storage OSystem , scan all pages to find the latest operation stored into the files, if you wish to skip this step to speed up data restore but decrease durability guarantees please set flag storage.checkLatestOperationId to the false [OLocalPaginatedStorage]
2021-03-20 20:11:18:637 INFO  Scanning of file ouser.pcl in storage OSystem (0 out of 67 files are scanned) [OLocalPaginatedStorage]Exception during storage data restore
java.lang.IllegalStateException: Id of WAL operation can not be duplicated
        at com.orientechnologies.orient.core.storage.impl.local.OAbstractPaginatedStorage.fetchNextOperationId(OAbstractPaginatedStorage.java:6502)
        at com.orientechnologies.orient.core.storage.impl.local.OAbstractPaginatedStorage.restoreFromBeginning(OAbstractPaginatedStorage.java:6221)
        at com.orientechnologies.orient.core.storage.impl.local.OAbstractPaginatedStorage.restoreFromWAL(OAbstractPaginatedStorage.java:6161)
        at com.orientechnologies.orient.core.storage.impl.local.OAbstractPaginatedStorage.recoverIfNeeded(OAbstractPaginatedStorage.java:5320)
        at com.orientechnologies.orient.core.storage.impl.local.OAbstractPaginatedStorage.open(OAbstractPaginatedStorage.java:401)
        at com.orientechnologies.orient.core.db.OrientDBEmbedded.getAndOpenStorage(OrientDBEmbedded.java:500)
        at com.orientechnologies.orient.core.db.OrientDBEmbedded.openNoAuthorization(OrientDBEmbedded.java:455)
        at com.orientechnologies.orient.core.db.OrientDBEmbedded.openNoAuthorization(OrientDBEmbedded.java:79)
        at com.orientechnologies.orient.server.OSystemDatabase.openSystemDatabase(OSystemDatabase.java:93)
        at com.orientechnologies.orient.server.OSystemDatabase.checkServerId(OSystemDatabase.java:169)
        at com.orientechnologies.orient.server.OSystemDatabase.init(OSystemDatabase.java:157)
        at com.orientechnologies.orient.server.OSystemDatabase.<init>(OSystemDatabase.java:50)
        at com.orientechnologies.orient.server.OServer.initSystemDatabase(OServer.java:1275)
        at com.orientechnologies.orient.server.OServer.activate(OServer.java:445)
```

**Comment by andrii0lomakin** [comment]
Hi, 
As I can see shutdown was not successful, that is seen because system start to scan WAL.
"2021-03-20 20:11:18:631 INFO  Storage OSystem , scan all pages to find the latest operation stored into the files, if you wish to skip this step to speed up data restore but decrease durability guarantees please set flag storage.checkLatestOperationId to the false [OLocalPaginatedStorage]
2021-03-20 20:11:18:637 INFO  Scanning of file ouser.pcl in storage OSystem (0 out of 67 files are scanned) [OLocalPaginatedStorage]Exception during storage "

Also, did you perform an upgrade of the library after the shutdown?

**Comment by PhantomYdn** [comment]
@laa , I agree that it looks strange that it started to scan, but the previous shutdown was successful per logs: at least no exceptions or error messages.

Originally db was created under 3.1.5, then library changed to 3.1.9 and it was working for some period of time under 3.1.9 till one day it started to fail as I posted.

**Comment by grimace** [comment]
I'm having the same issue, for 3 or versions now.   Most recently because the server was rebooted without shutting down OrientDB correctly,  tho there was no activity on the database during the time the server shutdown.  

I can no longer load the database with this error.  Cannot even load the DB in console for repair.  Is there any work around for this?  IMHO, the error is a bit obtuse.  Is there a better explanation of what the error means?

**Comment by jamieb22** [comment]
Andrey, I am not sure if you noticed yet, but there are three bug reports on the same issue. This issue appears bigger than just a corrupted database. Would you consider investigating this?

**Comment by andrii0lomakin** [comment]
Hi @jamieb22 . For sure. But now I am busy with commercial issues. So to fix it ASAP I would need a working test case. Otherwise, I can take it only after the fix of some issues which are submitted by our commercial clients.

**Comment by PhantomYdn** [comment]
Encountered again into the same issue on OrientDb 3.1.10.
[Orienteer_IdWalDuplicated.zip](https://github.com/orientechnologies/orientdb/files/6752320/Orienteer_IdWalDuplicated.zip)

I was doing nothing illegal: just debugging testcases for non OrientDB related code. Testcase itself closed DBproperly: was no issues, but next run DB was already broken.

**Comment by PhantomYdn** [comment]
@laa , please share some insights about this issue. Just got the same on one of our productions... Sad to loose half-day data:(

**Comment by jamieb22** [comment]
Downgrading to Orient DB 3.1.7 seems to help. Perhaps downgrade?

**Comment by andrii0lomakin** [comment]
Hi @PhantomYdn  @jamieb22  is right you can downgrade to 3.1.7 and then upgrade back to the version you needed. I will check your database once I have time.

**Comment by jamieb22** [comment]
Andrey, do you have any unit tests part of build process that test the upgrade and downgrade of Orient DB? Perhaps, the reason why so many people experience problems with upgrade & downgrade, is because the test coverage in this area needs to be more thorough.

**Comment by jamieb22** [comment]
Andrey. I see you added support for large files >2GB in WAL. Thank you for looking into this. I obviously do not know ODB internals, however, the attached database has the duplicate WAL issue observed when attempting to open it using v3.2.0 and there is no 2 GB WAL file present. Furthermore, if I delete the WAL files, the issue appears again. In fact,  no matter what I try, I cannot the db using 3.2.0. I need to downgrade to V3.1.7. 
[OSystem.zip](https://github.com/orientechnologies/orientdb/files/7124150/OSystem.zip)

**Comment by jamieb22** [comment]
Incidentally, I tried the latest SNAPSHOT on master branch to see if it would work. The duplicate WAL Id issue appears to be gone. I see a bunch of these messages:

2021-09-07 21:21:23:836 INFO  35 RidBags out of 226 are migrated ... PLEASE WAIT FOR COMPLETION ! [OSBTreeCollectionManagerShared]
2021-09-07 21:21:23:836 INFO  Migration of RidBag for cluster #77 is started ... PLEASE WAIT FOR COMPLETION ! [OSBTreeCollectionManagerShared]

Then after a while it started, to output an error. 

2021-09-07 21:17:07:509 INFO  Page size for WAL located in /mnt/malie is set to 4096 bytes. [CASDiskWriteAheadLog]Error during WAL writing
java.lang.IllegalStateException: There should be at least one record in the queue which has valid position
        at com.orientechnologies.orient.core.storage.impl.local.paginated.wal.cas.CASDiskWriteAheadLog.calculateRecordsLSNs(CASDiskWriteAheadLog.java:1488)
        at com.orientechnologies.orient.core.storage.impl.local.paginated.wal.cas.CASDiskWriteAheadLog.logMilestoneRecord(CASDiskWriteAheadLog.java:1517)
        at com.orientechnologies.orient.core.storage.impl.local.paginated.wal.cas.CASDiskWriteAheadLog.access$1200(CASDiskWriteAheadLog.java:85)
        at com.orientechnologies.orient.core.storage.impl.local.paginated.wal.cas.CASDiskWriteAheadLog$RecordsWriter.run(CASDiskWriteAheadLog.java:1742)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
Exception in thread 'OrientDB WAL Flush Task'

It just Outputs the following over and over. Since it is in the middle of a Ridbag conversion process, I am loathe to restart.

2021-09-07 21:25:29:442 SEVER Can not flush data because of issue during data write, java.util.concurrent.ExecutionException: java.lang.IllegalStateException: There should be at least one record in the queue which has valid position [OWOWCache$PeriodicFlushTask]

**Comment by jamieb22** [comment]
Then, I see the following outputted continually:

Caused by: com.orientechnologies.orient.core.exception.OStorageException: File global_collection_120.grb has already been added to the storage
        at com.orientechnologies.orient.core.storage.cache.local.OWOWCache.bookFileId(OWOWCache.java:624)
        at com.orientechnologies.orient.core.storage.impl.local.paginated.atomicoperations.OAtomicOperationBinaryTracking.addFile(OAtomicOperationBinaryTracking.java:342)
        at com.orientechnologies.orient.core.storage.impl.local.paginated.base.ODurableComponent.addFile(ODurableComponent.java:190)
        at com.orientechnologies.orient.core.storage.index.sbtreebonsai.global.btree.BTree.lambda$create$0(BTree.java:53)
        at com.orientechnologies.orient.core.storage.impl.local.paginated.atomicoperations.OAtomicOperationsManager.executeInsideComponentOperation(OAtomicOperationsManager.java:173)
        ... 17 more

**Comment by andrii0lomakin** [comment]
Hi, @jamieb22 could you provide your database so I will check those issues? Is it the database link on which you sent to me?

**Comment by andrii0lomakin** [comment]
@jamieb22 about the absence of the big WAL files, they are typically very short living and you likely will not notice them just by inspection of the directory. I am working on the fix for the 3.1 version too. Would be cool if you try it on your databases before I merge it.

**Comment by jamieb22** [comment]
I will certainly try it out on our databases. I will  also see if I can supply you one. I am getting lots of other errors with the latest SNAPSHOT master branch version. I am also seeing the below exception. I never got this before the upgrade and our code has not changed.

Caused by: com.orientechnologies.orient.core.exception.ODatabaseException: The database instance is not set in the current thread. Be sure to set it with: ODatabaseRecordThreadLocal.instance().set(db);
        at com.orientechnologies.orient.core.db.ODatabaseRecordThreadLocal.get(ODatabaseRecordThreadLocal.java:61)
        at com.orientechnologies.orient.core.sql.query.OSQLQuery.deserializeQueryParameters(OSQLQuery.java:147)
        at com.orientechnologies.orient.core.sql.query.OSQLQuery.queryFromStream(OSQLQuery.java:137)
        at com.orientechnologies.orient.core.sql.query.OSQLSynchQuery.queryFromStream(OSQLSynchQuery.java:147)
        at com.orientechnologies.orient.core.sql.query.OSQLQuery.fromStream(OSQLQuery.java:109)
        at com.orientechnologies.orient.core.serialization.serializer.stream.OStreamSerializerAnyStreamable.fromStream(OStreamSerializerAnyStreamable.java:86)
        ... 4 more

**Comment by jamieb22** [comment]
FYI: This duplicate WAL Id iissue is still present in 3.2.2. To reproduce the issue, sample database can be downloaded from #9542.

**Comment by andrii0lomakin** [comment]
Hi, @jamieb22 that is because the fix was reverted. I am working on the final touches and will send you an update once it is done.

**Comment by andrii0lomakin** [comment]
Fixed both in 3.1 and 3.2
